util/network-devp2p/src/discovery.rs

```diff
@@ -529,38 +531,40 @@ impl<'a> Discovery<'a> {
 				expected_node
 			},
 			Entry::Vacant(_) => {
-				debug!(target: "discovery", "Got unexpected Pong from {:?} ; request not found", &from);
 				None
 			},
 		};
 
 		// If the Node couldn't be found from its ID,
 		// Try to find it in from the hash that was sent back
-		// if expected_node.is_none() {
-		// 	self.in_flight_pings.retain(|_, ping_request| {
-		// 		if ping_request.echo_hash == echo_hash || ping_request.deprecated_echo_hash == echo_hash {
-		// 			debug!(target: "discovery",
-		// 				"Found corresponding request expected={:x}@{} ; found={:x}@{}",
-		// 				node_id, from,
-		// 				ping_request.node.id, ping_request.node.endpoint.address
-		// 			);
-		// 			expected_node = Some(NodeEntry {
-		// 				id: *node_id,
-		// 				endpoint: NodeEndpoint {
-		// 					udp_port: ping_request.node.endpoint.udp_port,
-		// 					address: from.clone(),
-		// 				}
-		// 			});
-		// 			false
-		// 		} else {
-		// 			true
-		// 		}
-		// 	});
-		// }
+		if expected_node.is_none() {
+			let mut found = false;
+			self.in_flight_pings.retain(|_, ping_request| {
```

Comment:
When would this be the case? Iterating through the whole `in_flight_pings` map on every unexpected PONG doesn't seem like a good idea to me.

If there's a good reason for this, the map might as well be keyed on the ping hash (except that it would be harder to maintain compatibility with the `deprecated_echo_hash` thing).

Reply:
So this is to cater for the fact that the node answering your Ping might not have the NodeID you've been told. Another mechanism would be to Ping back any unknown received Pong ; I can try that if it sounds better.

Reply:
If the NodeID is wrong, then the address (NodeID + Network Endpoint) is wrong, so I think it's fine to just drop the PONG.

Consider also that the sender of the PONG will likely PING back to us after receiving the PING, from which we can get the correct NodeID, so this unlikely case should resolve itself anyway.

Reply:
Okay I'll remove it from now. Note that Geth doesn't Ping back an received Ping, it just adds it to their NodeTable .